fn invalid_emoji_usages() {
    let arrowâ†”ï¸ = "basic emoji"; // { dg-error "" "" { target *-*-* } }
    let planetğŸª = "basic emoji"; // { dg-error "" "" { target *-*-* } }
    let wirelessğŸ›œ = "basic emoji"; // { dg-error "" "" { target *-*-* } }
    // FIXME
    let key1ï¸âƒ£ = "keycap sequence"; // { dg-error "" "" { target *-*-* } }
// { dg-warning "" "" { target *-*-* } .-1 }
    let flagğŸ‡ºğŸ‡³ = "flag sequence"; // { dg-error "" "" { target *-*-* } }
    let walesğŸ´ = "tag sequence"; // { dg-error "" "" { target *-*-* } }
    let foldedğŸ™ğŸ¿ = "modifier sequence"; // { dg-error "" "" { target *-*-* } }
}

fn main() {
    invalid_emoji_usages();
}

