struct ABigğŸ‘©â€ğŸ‘©â€ğŸ‘§â€ğŸ‘§Family; // { dg-error "" "" { target *-*-* } }
struct ğŸ‘€; // { dg-error "" "" { target *-*-* } }
impl ğŸ‘€ {
    fn full_of_âœ¨() -> ğŸ‘€ { // { dg-error "" "" { target *-*-* } }
        ğŸ‘€
    }
}
fn i_like_to_ğŸ˜…_a_lot() -> ğŸ‘€ { // { dg-error "" "" { target *-*-* } }
    ğŸ‘€::full_ofâœ¨() // { dg-error ".E0599." "" { target *-*-* } }
// { dg-error ".E0599." "" { target *-*-* } .-1 }
}
fn main() {
    let _ = i_like_to_ğŸ˜„_a_lot() â– 4; // { dg-error ".E0425." "" { target *-*-* } }
// { dg-error ".E0425." "" { target *-*-* } .-1 }
// { dg-error ".E0425." "" { target *-*-* } .-2 }

    let ğŸ¦€ = 1;// { dg-error "" "" { target *-*-* } }
    dbg!(ğŸ¦€);
}

